lib/versioning/regex/index.ts

```diff
@@ -0,0 +1,151 @@
+  // never try to update to a version with a different architecture. Other than
+  // for juding compatibility (exact string match), the architecture is ignored
+  architecture: string;
+}
+
+// convenience method for passing a Version object into any semver.* method.
+function asSemver(version: Version): string {
+  let vstring = `${version.major}.${version.minor}.${version.patch}`;
+  if (typeof version.prerelease !== 'undefined') {
+    vstring += `-${version.prerelease}`;
+  }
+  return vstring;
+}
+
+export const api: VersioningApi = {
+  // config is expected to be overridden by a user-specified RegExp value
+  // sample values:
+  //
+  // * emulates the "semver" configuration:
+  //   RegExp('^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(-(?<prerelease>.*))?$')
+  // * emulates the "docker" configuration:
+  //   RegExp('^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(-(?<architecture>.*))?$')
+  // * matches the versoining scheme used by the Python images on DockerHub:
+  //   RegExp('^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(?<prerelease>[^.-]+)?(-(?<architecture>.*))?$');
+  //
+  // defaults to a simple semver scheme
+  config: null,
+
+  // TODO: the linter expects method shorthands for all these methods, but
+  // those don't seem to bind `this`. What's the middle-ground?
+  configure: function(config) {
+    // without at least one of {major, minor, patch} specified in the regex,
+    // this versioner will not work properly
+    // TODO: there's got to be a better way to do this
+    if (config.includes('<major>') + config.includes('<minor>') + config.includes('<patch>') === 0) {
+      logger.fatal({ config }, 'Invalid versionConfig for versionScheme regex');
+      return;
+    }
+
+    // TODO: should we validate the user has not added extra unsupported
+    // capture groups?
+
+    this.config = RegExp(config);
+  },
+
+  parse: function(version: string): Version {
+    const match = version.match(this.config);
+    if (match === null) {
+      // TODO: what are we expected to return for invalid functions?
+      // return null;
+      return {
+        major: 0,
+        minor: 0,
+        patch: 0,
+        prerelease: undefined,
+        architecture: 'invalid',
+      };
+    }
+
+    const groups = match.groups;
+    return {
+      major: (typeof groups.major === 'undefined' ? 0 : Number(groups.major)),
+      minor: (typeof groups.minor === 'undefined' ? 0 : Number(groups.minor)),
+      patch: (typeof groups.patch === 'undefined' ? 0 : Number(groups.patch)),
+      prerelease: groups.prerelease,
+      architecture: groups.architecture
+    };
+  },
+
+  isCompatible: function(version: string, range: string): boolean {
+    return this.parse(version).architecture === this.parse(range).architecture;
+  },
+  isSingleVersion: function(version: string): string | boolean | null {
+    // TODO: whats the difference between isSingleVersion, isValid, and isVersion?
```

Comment:
@ contributors: could I get a rundown of the difference between `isSingleVersion`, `isVersion`, and `isValid`? My impression is that for this versioner they'd all be the same (eg. any version that can be parsed as a version is inherently valid and not a range), but I got a bit lost following all the code paths here.

Reply:
Copying from a response from @rarkins (thanks!)

> `isSingleVersion` exists because versioning schemes have different syntax. e.g. for python you need `==1.0.0` while for other managers `1.0.0` means "1.0.0 or above", etc. So it returns true if the provided string represents a pinned/single version
>
> in this regex case I would not support ranges so isSingleVersion/isVersion/isValid would probably all return the same answer every time
>
> `isVersion` means the string is a version (e.g. `1.0.0`) while isValid can include ranges too (e.g. >= `1.0.0`)
> 
> `isValid` should return false for something like `foo`. Ideally your `isVersion` and `isSingleVersion` functions call to `isValid` first though perhaps, in order to be defensive
> 
> ideally the other functions would not get called by the application layer if `isValid` is false, but while writing the versioning API we can't be sure of that